src/v2i-hub/ERVCloudForwardingPlugin/src/ERVCloudForwardingPlugin.cpp

```diff
@@ -0,0 +1,214 @@
+#include "ERVCloudForwardingPlugin.h"
+
+namespace ERVCloudForwardingPlugin
+{
+    ERVCloudForwardingPlugin::ERVCloudForwardingPlugin(const string &name) : PluginClient(name)
+    {
+        UpdateConfigSettings();
+        std::lock_guard<mutex> lock(_cfgLock);
+        AddMessageFilter<BsmMessage>(this, &ERVCloudForwardingPlugin::handleBSM);
+        // Subscribe to all messages specified by the filters above.
+        SubscribeToMessages();
+
+        std::thread webBSM_t(&ERVCloudForwardingPlugin::StartBSMWebService, this);
+        webBSM_t.detach();
+
+        // Send RSU location to cloud
+        RegisterRSULocation();
+    }
+
+    void ERVCloudForwardingPlugin::handleBSM(BsmMessage &msg, routeable_message &routableMsg)
+    {
+        // Check if the BSM is broadcast by an ERV (Emergency Response Vehicle)
+        if (ERVCloudForwardingWorker::IsBSMFromERV(msg))
+        {
+            // Construct the ERV BSM and forward it to the cloud.
+            auto xml_str = ERVCloudForwardingWorker::constructERVBSMRequest(msg);
+            PLOG(logINFO) << "Forward ERV BSM to cloud: " << xml_str << endl;
+            CloudSendAsync(xml_str, _CLOUDURL, _CLOUDBSMREQ, _POSTMETHOD);
+        }
+        else
+        {
+            // If BSM is not from ERV, print debug log
+            PLOG(logDEBUG) << "Incoming BSM is not from Emergency Response Vehicle (ERV)." << endl;
+        }
+    }
+
+    void ERVCloudForwardingPlugin::RegisterRSULocation()
+    {
+        try
+        {
+            PLOG(logINFO) << "Create SNMP Client to connect to RSU. RSU IP:" << _rsuIp << ",\tRSU Port:" << _snmpPort << ",\tSecurity Name:" << _securityUser << ",\tAuthentication Passphrase: " << _authPassPhrase << endl;
+            auto snmpClient = std::make_shared<SNMPClient>(_rsuIp, _snmpPort, _securityUser, _authPassPhrase);
+            auto gps_sentense = snmpClient->SNMPGet(_GPSOID);
+            auto gps_map = ERVCloudForwardingWorker::ParseGPS(gps_sentense);
+            long latitude = 0;
+            long longitude = 0;
+            for (auto itr = gps_map.begin(); itr != gps_map.end(); itr++)
+            {
+                latitude = itr->first;
+                longitude = itr->second;
+            }
+
+            if (latitude == 0 || longitude == 0)
+            {
+                PLOG(logERROR) << "Invalid latitude and longitude. Cannot register RSU location." << endl;
+                return;
```

Comment:
Do we want to throw exceptions here instead of returning null?

Reply:
I prefer to log the error and not end the execution by throwing the exception.

Reply:
You can catch the error in the calling function and handle it in some specific way. I think like this, the calling function would have no indication of whether an error occurred or not so it would not be possible to write any custom handling code for these cases.

Reply:
yeah, I think this is the calling function itself. The exception usually throw from other class other than inside the plugin class itself.